shared/src/main/java/uk/gov/di/authentication/shared/services/RedisConnectionService.java

```diff
@@ -40,63 +43,48 @@ public RedisConnectionService(ConfigurationService configurationService) {
-        } catch (Exception e) {
-            LOGGER.error(REDIS_CONNECTION_ERROR, e);
-            throw new RedisConnectionException(REDIS_CONNECTION_ERROR, e);
-        }
+    @FunctionalInterface
+    private interface RedisFunction<T> {
+        T getResult(RedisCommands<String, String> commands);
     }
 
-    public boolean keyExists(String key) {
+    private <T> T executeCommand(RedisFunction<T> callable) {
         try (StatefulRedisConnection<String, String> connection = pool.borrowObject()) {
-            return (connection.sync().exists(key) == 1);
+            return callable.getResult(connection.sync());
         } catch (Exception e) {
-            LOGGER.error(REDIS_CONNECTION_ERROR, e);
             throw new RedisConnectionException(REDIS_CONNECTION_ERROR, e);
         }
     }
 
-    public String getValue(String key) {
-        try (StatefulRedisConnection<String, String> connection = pool.borrowObject()) {
-            return connection.sync().get(key);
-        } catch (Exception e) {
-            LOGGER.error(REDIS_CONNECTION_ERROR, e);
-            throw new RedisConnectionException(REDIS_CONNECTION_ERROR, e);
-        }
+    public void saveWithExpiry(final String key, final String value, final long expiry) {
+        executeCommand(commands -> commands.setex(key, expiry, value));
     }
 
-    public long deleteValue(String key) {
-        try (StatefulRedisConnection<String, String> connection = pool.borrowObject()) {
-            return connection.sync().del(key);
-        } catch (Exception e) {
-            LOGGER.error(REDIS_CONNECTION_ERROR, e);
-            throw new RedisConnectionException(REDIS_CONNECTION_ERROR, e);
-        }
+    public boolean keyExists(final String key) {
+        return executeCommand(commands -> commands.exists(key) == 1);
     }
 
-    public String popValue(String key) {
-        try (StatefulRedisConnection<String, String> connection = pool.borrowObject()) {
-            RedisCommands<String, String> commands = connection.sync();
-            commands.multi();
-            commands.get(key);
-            commands.del(key);
-            TransactionResult result = commands.exec();
-            return result.get(0);
-        } catch (Exception e) {
-            LOGGER.error(REDIS_CONNECTION_ERROR, e);
-            throw new RedisConnectionException(REDIS_CONNECTION_ERROR, e);
-        }
+    public String getValue(final String key) {
+        return executeCommand(commands -> commands.get(key));
+    }
+
+    public long deleteValue(final String key) {
+        return executeCommand(commands -> commands.del(key));
+    }
+
+    public String popValue(final String key) {
+        return executeCommand(
+                commands -> {
+                    commands.multi();
+                    commands.get(key);
+                    commands.del(key);
+                    TransactionResult result = commands.exec();
+                    return result.get(0);
```

Comment:
We could probably refactor this to use `commands.getdel(key)`?

https://redis.io/commands/getdel/